web/src/buddy_list.js

```diff
@@ -55,16 +87,259 @@ class BuddyListConf {
+                                        "{N, plural, one {# subscriber} other {# subscribers}}",
+                                },
+                                {N: subscriber_count},
+                            );
+                        } else {
+                            tooltip_text = $t(
+                                {
+                                    defaultMessage:
+                                        "{N, plural, one {# participant} other {# participants}}",
+                                },
+                                {N: subscriber_count},
+                            );
+                        }
+                    } else {
+                        const total_user_count = people.get_active_human_count();
+                        const other_users_count = total_user_count - subscriber_count;
+                        tooltip_text = $t(
+                            {
+                                defaultMessage:
+                                    "{N, plural, one {# other user} other {# other users}}",
+                            },
+                            {N: other_users_count},
+                        );
+                    }
+                    instance.setContent(tooltip_text);
+                },
+                onHidden(instance) {
+                    instance.destroy();
+                },
+                appendTo: () => document.body,
+            });
+        });
+        this.tooltips_initialized = true;
+    }
+
+    set_render_data() {
+        const current_sub = narrow_state.stream_sub();
+        const pm_ids_set = narrow_state.pm_ids_set();
+
+        // If we have only "other users" and aren't in a stream/DM view
+        // then we don't show section headers and only show one untitled
+        // section.
+        const hide_headers = this.should_hide_headers(current_sub, pm_ids_set);
+        const subscriber_count = this.total_subscriber_count(current_sub, pm_ids_set);
+        const total_user_count = people.get_active_human_count();
+        const other_users_count = total_user_count - subscriber_count;
+        const has_inactive_users_matching_view =
+            subscriber_count > this.users_matching_view_ids.length;
+        const has_inactive_other_users = other_users_count > this.other_user_ids.length;
+
+        this.render_data = {
+            current_sub,
+            hide_headers,
+            subscriber_count,
+            other_users_count,
+            has_inactive_users_matching_view,
+            has_inactive_other_users,
+            total_user_count,
+        };
+    }
 
     populate(opts) {
         this.render_count = 0;
-        this.$container.empty();
+        this.$users_matching_view_container.empty();
+        this.users_matching_view_ids = [];
+        this.$other_users_container.empty();
+        this.other_user_ids = [];
+
+        // Reset data to be relevant for this current view.
+        this.set_render_data();
```

Comment:
So `this.set_render_data` seems to inspect values like `users_matching_view_ids`, both directly and indirectly (via `should_hide_headers`); what's the thinking of calling this with stale/bogus data at this point?

Reply:
Oh shoot yeah, it should be calculating `should_hide_headers` here. I think everything else doesn't use the bogus data. The way to guarantee this would be to pull the data calculation fully out of the BuddyList class, along with `total_subscriber_count`. I'll do that now.

Reply:
Pushing the last commit with that change. I'll come back to look at your other code review comments later (maybe tomorrow), gotta head out for a bit now

Reply:
That refactor looks like an improvement.